rosidl_generator_cpp/resource/msg__struct.hpp.em

```diff
@@ -30,13 +32,18 @@ msvc_common_macros = ('DELETE', 'ERROR', 'NO_ERROR')
 @# Collect necessary include directives for all members
 @{
 from collections import OrderedDict
-from rosidl_cmake import convert_camel_case_to_lower_case_underscore
+from rosidl_pycommon import convert_camel_case_to_lower_case_underscore
 includes = OrderedDict()
 for member in message.structure.members:
     type_ = member.type
     if isinstance(type_, AbstractNestedType):
         type_ = type_.value_type
     if isinstance(type_, NamespacedType):
+        if (
+            type_.name.endswith(SERVICE_REQUEST_MESSAGE_SUFFIX) or
+            type_.name.endswith(SERVICE_RESPONSE_MESSAGE_SUFFIX)
+        ):
+            continue
```

Comment:
To assist with review, can you explain why we should check for service suffixes here and ignore them?

Reply:
Service suffixes are checked here in order to avoid generating `#include`s for generated service request, response, and event messages [`__trait`](https://github.com/ros2/rosidl/blob/46753d73dd1fc6e9208330ee8e6f340db72157de/rosidl_generator_cpp/resource/msg__traits.hpp.em#L35-L34)s and `__struct`s. This is because, unlike how a message that embeds another message defined "traditionally" i.e. creating a .msg file would need to `#include` the correspoinding traits/struct/etc, the service request/response/events already live together in the produced traits/struct/etc.

This is also how it is done with action messages in the same file.

Reply:
It make sense that we have a special case for these service types, but why not handle them similar to what we're doing for actions. Instead of including nothing, we include the header that actually contains the declaration (e.g. at L52, we lop off the action suffix).

TBH, I'm not sure under what case we have a message that contains a nested action type, but I guess it doesn't hurt to allow it here.

Reply:
Yeah, that'd make sense in the case of nested action or service types (which are still unsupported outside of injecting them into the rosidl pipeline as we're doing in this PR). But for currently supported usecases these `#include`s are referencing the file itself and it's a little circular (though harmless because of the header guards) For example `example_interfaces/AddTwoInts.srv` would create a `rosidl_generator_cpp/detail/add_two_ints__struct.hpp` which includes the `_Request`, `_Response`, and `Event` messages -- the declarations for which already live inside `rosidl_generator_cpp/detail/add_two_ints__struct.hpp`, and the same would apply to actions. I _think_ we could even just change that entire block to

```cpp
        if (
            type_.name.endswith(SERVICE_REQUEST_MESSAGE_SUFFIX) or
            type_.name.endswith(SERVICE_RESPONSE_MESSAGE_SUFFIX) or
            type_.name.endswith(ACTION_GOAL_SUFFIX) or
            type_.name.endswith(ACTION_RESULT_SUFFIX) or
            type_.name.endswith(ACTION_FEEDBACK_SUFFIX)
        ):
            continue
```

But if that nesting is something we want then it's just 

```suggestion
            typename = type_.name.rsplit('_', 1)[0]
```

Either should be fine

Reply:
I see. The circular includes _are_ odd (I recall being confused because of them when looking at generated code). Maybe it is best to just ignore them like you suggest.

Reply:
This is to avoid recursive inclusion of the same file, it isn't a problem but it looks weird.
Same change needs to be applied to `rosidl_generator_c`.

Reply:
I have done some modifications of this in https://github.com/ros2/rosidl/pull/700/commits/8e66f38e476d27ebc4cf5b1a9932e5620c010cb7